components/core/src/reducer/Operator.hpp

```diff
@@ -0,0 +1,69 @@
+#ifndef REDUCER_OPERATOR_HPP
+#define REDUCER_OPERATOR_HPP
+
+#include <memory>
+#include <set>
+#include <utility>
+
+#include "RecordGroup.hpp"
+#include "RecordGroupIterator.hpp"
+
+namespace reducer {
+enum class OperatorType : uint8_t {
+    Map,
+    GroupBy,
+    Reduce
+};
+
+enum class OperatorResultCardinality : uint8_t {
+    One,
+    Subset,
+    Input
+};
+
+/**
+ * Class implementing a generic Operator that operates on RecordGroup objects.
+ *
+ * Operators implement methods to accept inter-stage RecordGroups and intra-stage RecordGroups.
+ * In the former case, the operator acts like a combiner, combining records from a previous pipeline
+ * stage. In the latter case, the operator acts like a reducer, reducing pre-aggregated results from
+ * within a pipeline stage.
+ *
+ * Operators can be of type GroupBy, Map, or Reduce, though currently no Map type Operators are
+ * implemented.
+ */
+class Operator {
+public:
+    virtual ~Operator() = default;
+
+    virtual void
+    push_intra_stage_record_group(GroupTags const& tags, ConstRecordIterator& record_it)
+            = 0;
+    virtual void
+    push_inter_stage_record_group(GroupTags const& tags, ConstRecordIterator& record_it)
+            = 0;
```

Comment:
How about `reduce_record_group` and `combine_record_group`?

Reply:
I guess we'd have to make a convention where map/groupby type operators are implemented in the combine method or something, but I'm not necessarily opposed.

Reply:
I think I'm going to leave this as is for now, and we can rename things if/when we decide to make bigger changes to this interface